compat/c_layer.go

```diff
@@ -1,48 +1,159 @@
 package main
 
-import "fmt"
-import "github.com/google/go-jsonnet"
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
 
-// #cgo CXXFLAGS: -std=c++11 -Wall
-import "C"
+	"github.com/google/go-jsonnet"
 
-//export test
-func test() {
-	fmt.Printf("hello")
+	// #cgo CXXFLAGS: -std=c++11 -Wall
+	// #include "internal.h"
+	"C"
+)
+
+const maxID = 1000
+
+type vm struct {
+	*jsonnet.VM
+	importer *jsonnet.FileImporter
 }
 
-//export run_jsonnet
-func run_jsonnet(code *C.char) {
-	s := C.GoString(code)
-	vm := jsonnet.MakeVM()
-	out, err := vm.EvaluateSnippet("c", s)
-	if err != nil {
-		fmt.Println(err)
+// Because of Go GC, there are restrictions on keeping Go pointers in C.
+// We cannot just pass *jsonnet.VM to C. So instead we use "handle" structs in C
+// which refer to JsonnetVM by a numeric id.
+// The way it is implemented below is simple and has low overhead, but requires us to keep
+// a list of used IDs. This results in a permanent "leak". I don't expect it to ever
+// become a problem.
+var VMs = []*vm{}
+var freedIDs = []uint32{}
+
+var versionString *C.char
+
+//export jsonnet_version
+func jsonnet_version() *C.char {
+	if versionString == nil {
+		version := jsonnet.Version() + " (go-jsonnet)"
+		versionString = C.CString(version)
+	}
+	return versionString
+}
+
+//export jsonnet_make
+func jsonnet_make() *C.struct_JsonnetVm {
+	var id uint32
+
+	newVM := &vm{jsonnet.MakeVM(), &jsonnet.FileImporter{}}
+	newVM.Importer(newVM.importer)
+
+	if len(freedIDs) > 0 {
+		id, freedIDs = freedIDs[len(freedIDs)-1], freedIDs[:len(freedIDs)-1]
+		VMs[id-1] = newVM
 	} else {
-		fmt.Println(out)
+		id = uint32(len(VMs) + 1)
+		if id > maxID {
+			fmt.Fprintf(os.Stderr, "Maximum number of constructed Jsonnet VMs exceeded (%d)\n", maxID)
+			os.Exit(1)
+		}
+		VMs = append(VMs, newVM)
 	}
+	addr := C.jsonnet_internal_make_vm_with_id(C.uint32_t(id))
+	return addr
 }
 
-//export jsonnet_evaluate_snippet2
-func jsonnet_evaluate_snippet2(filename *C.char, code *C.char, e *C.int) *C.char {
-	f := C.GoString(filename)
-	fmt.Println(code)
-	s := C.GoString(code)
-	vm := jsonnet.MakeVM()
-	fmt.Println(s)
-	out, err := vm.EvaluateSnippet(f, s)
+//export jsonnet_destroy
+func jsonnet_destroy(vmRef *C.struct_JsonnetVm) {
+	VMs[vmRef.id-1] = nil
+	freedIDs = append(freedIDs, uint32(vmRef.id))
+	C.jsonnet_internal_free(vmRef)
```

Comment:
I wonder if it's better to just declare `void free(void *ptr);` in internal.h?  Is there an explicit best practice for this?

Reply:
Looks like you can just do C.free(unsafe.Pointer(vmRef))

Reply:
...actually you're creating with `new` so you have to clean it up with `delete`, so you do need jsonnet_internal_free but it shouldn't call `free` in its implementation.

Reply:
done

I think I could get rid of these functions by just using `malloc`/`free` on Go side, so I can also do that if that's better. Sizeof is a bit ugly on Go side, but it should work.